src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -89,57 +166,79 @@ public static SourceLocation Add(SourceLocation left, SourceLocation right)
             else
             {
                 return new SourceLocation(
+                    left.FilePath,
                     left.AbsoluteIndex + right.AbsoluteIndex,
                     left.LineIndex + right.LineIndex,
                     left.CharacterIndex + right.CharacterIndex);
             }
         }
 
-        public static SourceLocation Subtract(SourceLocation left, SourceLocation right)
+        /// <summary>
+        /// Subtracts two <see cref="SourceLocation"/>s.
+        /// </summary>
+        /// <param name="left">The left operand.</param>
+        /// <param name="right">The right operand.</param>
+        /// <returns>A <see cref="SourceLocation"/> that is the difference of the left and right operands.</returns>
+        /// <exception cref="ArgumentException">if the <see cref="FilePath"/> of the left and right operands
+        /// are different.</exception>
+        public static SourceLocation operator -(SourceLocation left, SourceLocation right)
         {
+            if (!string.Equals(left.FilePath, right.FilePath, StringComparison.Ordinal))
+            {
+                throw new ArgumentException(
+                    RazorResources.FormatSourceLocationFilePathDoesNotMatch(nameof(SourceLocation)),
+                    nameof(right));
+            }
+
             return new SourceLocation(
+                left.FilePath,
```

Comment:
I'm neutral about this - subtraction isn't used anywhere in the codebase. Additions are used where the right operand is a relative location and it's used in calculating the offset from document start. So it sort of makes sense to preserve the document location in that case.